internal/keepers/ocr.go

```diff
@@ -53,50 +55,73 @@ func (k *keepers) Query(_ context.Context, _ types.ReportTimestamp) (types.Query
 // Observation implements the types.ReportingPlugin interface in OCR2. This method samples a set
 // of upkeeps available in and UpkeepService and produces an observation containing upkeeps that
 // need to be executed.
-func (k *keepers) Observation(ctx context.Context, rt types.ReportTimestamp, _ types.Query) (types.Observation, error) {
-	lCtx := newOcrLogContext(rt)
-	ctx = context.WithValue(ctx, ocrLogContextKey{}, lCtx)
+func (k *keepers) Observation(ctx context.Context, reportTimestamp types.ReportTimestamp, _ types.Query) (types.Observation, error) {
+	lCtx := newOcrLogContext(reportTimestamp)
 
-	blockKey, results, err := k.service.SampleUpkeeps(ctx, k.filter.Filter())
-	if err != nil {
-		return nil, fmt.Errorf("%w: failed to sample upkeeps for observation: %s", err, lCtx)
+	if len(k.observers) == 0 {
+		return nil, fmt.Errorf("empty observer list: %s", lCtx)
 	}
 
-	// keyList produces a sorted result so the following reduction of keys
-	// should be more uniform for all nodes
-	keys := keyList(filterUpkeeps(results, ktypes.Eligible))
+	allIDs := make([]ktypes.UpkeepIdentifier, 0)
+	var blocks []*big.Int
 
-	obs := &chain.UpkeepObservation{
-		BlockKey: chain.BlockKey(blockKey.String()),
-	}
+	for _, observer := range k.observers {
+		block, ids, err := observer.Observe()
+		if err != nil {
+			return nil, fmt.Errorf("%w: failed to sample upkeeps for observation: %s", err, lCtx)
+		}
+
+		if block == nil {
+			k.logger.Printf("observed %d upkeep IDs with nil block", len(allIDs))
```

Comment:
nit. should this be an error condition?

Reply:
@EasterTheBunny wdyt? Should we return an error at this point? I think I explicitly added this nil check purely for logging purposes, so I can just drop it completely and either always assume we won't get a nil block, OR only append IDs if the block is non nil, OR just return an error on a nil block?